### ISSUE TITLE
test: fs_event_watch_dir fails on OSX

### ISSUE THREAD

**Issue by hems** [libuv/libuv#30]
Thank you for your awesome work guys. :+1: 

Building @ Darwin, got this error :

````
deps/libuv [ sh autogen.sh;./configure;make;make check;make install                                                                    master ] 2:54 pm
+ glibtoolize
glibtoolize: putting auxiliary files in `.'.
glibtoolize: linking file `./ltmain.sh'
glibtoolize: putting macros in AC_CONFIG_MACRO_DIR, `m4'.
glibtoolize: linking file `m4/libtool.m4'
glibtoolize: linking file `m4/ltoptions.m4'
glibtoolize: linking file `m4/ltsugar.m4'
glibtoolize: linking file `m4/ltversion.m4'
glibtoolize: linking file `m4/lt~obsolete.m4'
+ aclocal -I m4
+ autoconf
+ automake --add-missing --copy
configure.ac:29: installing './ar-lib'
configure.ac:25: installing './compile'
configure.ac:22: installing './config.guess'
configure.ac:22: installing './config.sub'
configure.ac:21: installing './install-sh'
configure.ac:21: installing './missing'
Makefile.am: installing './depcomp'
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... ./install-sh -c -d
checking for gawk... no
checking for mawk... no
checking for nawk... no
checking for awk... awk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking build system type... x86_64-apple-darwin13.4.0
checking host system type... x86_64-apple-darwin13.4.0
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking whether gcc understands -c and -o together... yes
checking for style of include used by make... GNU
checking dependency style of gcc... gcc3
checking if gcc supports -Wno-dollar-in-identifier-extension flag... yes
checking for ar... ar
checking the archiver (ar) interface... ar
checking whether make supports nested variables... (cached) yes
checking how to print strings... printf
checking for a sed that does not truncate output... /usr/bin/sed
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for fgrep... /usr/bin/grep -F
checking for ld used by gcc... /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld
checking if the linker (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld) is GNU ld... no
checking for BSD- or MS-compatible name lister (nm)... /usr/bin/nm
checking the name lister (/usr/bin/nm) interface... BSD nm
checking whether ln -s works... yes
checking the maximum length of command line arguments... 196608
checking whether the shell understands some XSI constructs... yes
checking whether the shell understands "+="... yes
checking how to convert x86_64-apple-darwin13.4.0 file names to x86_64-apple-darwin13.4.0 format... func_convert_file_noop
checking how to convert x86_64-apple-darwin13.4.0 file names to toolchain format... func_convert_file_noop
checking for /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld option to reload object files... -r
checking for objdump... no
checking how to recognize dependent libraries... pass_all
checking for dlltool... no
checking how to associate runtime and link libraries... printf %s\n
checking for archiver @FILE support... no
checking for strip... strip
checking for ranlib... ranlib
checking command to parse /usr/bin/nm output from gcc object... ok
checking for sysroot... no
checking for mt... no
checking if : is a manifest tool... no
checking for dsymutil... dsymutil
checking for nmedit... nmedit
checking for lipo... lipo
checking for otool... otool
checking for otool64... no
checking for -single_module linker flag... yes
checking for -exported_symbols_list linker flag... yes
checking for -force_load linker flag... yes
checking how to run the C preprocessor... gcc -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking for dlfcn.h... yes
checking for objdir... .libs
checking if gcc supports -fno-rtti -fno-exceptions... yes
checking for gcc option to produce PIC... -fno-common -DPIC
checking if gcc PIC flag -fno-common -DPIC works... yes
checking if gcc static flag -static works... no
checking if gcc supports -c -o file.o... yes
checking if gcc supports -c -o file.o... (cached) yes
checking whether the gcc linker (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld) supports shared libraries... yes
checking dynamic linker characteristics... darwin13.4.0 dyld
checking how to hardcode library paths into programs... immediate
checking whether stripping libraries is possible... yes
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... yes
checking for dlopen in -ldl... yes
checking for kstat_lookup in -lkstat... no
checking for kvm_open in -lkvm... no
checking for gethostbyname in -lnsl... no
checking for perfstat_cpu in -lperfstat... no
checking for pthread_mutex_init in -lpthread... yes
checking for clock_gettime in -lrt... no
checking for sendfile in -lsendfile... no
checking for socket in -lsocket... no
checking for special C compiler options needed for large files... no
checking for _FILE_OFFSET_BITS value needed for large files... no
checking for pkg-config... yes
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating libuv.pc
config.status: creating Makefile
config.status: executing depfiles commands
config.status: executing libtool commands
  CC       src/libuv_la-fs-poll.lo
  CC       src/libuv_la-inet.lo
  CC       src/libuv_la-threadpool.lo
  CC       src/libuv_la-uv-common.lo
  CC       src/libuv_la-version.lo
  CC       src/unix/libuv_la-async.lo
  CC       src/unix/libuv_la-core.lo
  CC       src/unix/libuv_la-dl.lo
  CC       src/unix/libuv_la-fs.lo
  CC       src/unix/libuv_la-getaddrinfo.lo
  CC       src/unix/libuv_la-getnameinfo.lo
  CC       src/unix/libuv_la-loop-watcher.lo
  CC       src/unix/libuv_la-loop.lo
  CC       src/unix/libuv_la-pipe.lo
  CC       src/unix/libuv_la-poll.lo
  CC       src/unix/libuv_la-process.lo
  CC       src/unix/libuv_la-signal.lo
  CC       src/unix/libuv_la-stream.lo
src/unix/stream.c:1054:19: warning: variable length array folded to constant array as an extension [-Wgnu-folding-constant]
  char cmsg_space[CMSG_SPACE(UV__CMSG_FD_SIZE)];
                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/sys/socket.h:517:24: note: expanded from macro 'CMSG_SPACE'
#define CMSG_SPACE(l)           (__DARWIN_ALIGN32(sizeof(struct cmsghdr)) + __DARWIN_ALIGN32(l))
                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/unix/stream.c:1054:19: note: cast that performs the conversions of a reinterpret_cast is not allowed in a constant expression
/usr/include/sys/socket.h:517:25: note: expanded from macro 'CMSG_SPACE'
#define CMSG_SPACE(l)           (__DARWIN_ALIGN32(sizeof(struct cmsghdr)) + __DARWIN_ALIGN32(l))
                                 ^
/usr/include/i386/_param.h:43:42: note: expanded from macro '__DARWIN_ALIGN32'
#define       __DARWIN_ALIGN32(p)       ((__darwin_size_t)((char *)(__darwin_size_t)(p) + __DARWIN_ALIGNBYTES32) &~ __DARWIN_ALIGNBYTES32)
                                         ^
1 warning generated.
  CC       src/unix/libuv_la-tcp.lo
  CC       src/unix/libuv_la-thread.lo
  CC       src/unix/libuv_la-timer.lo
  CC       src/unix/libuv_la-tty.lo
  CC       src/unix/libuv_la-udp.lo
  CC       src/unix/libuv_la-darwin.lo
  CC       src/unix/libuv_la-darwin-proctitle.lo
  CC       src/unix/libuv_la-fsevents.lo
  CC       src/unix/libuv_la-kqueue.lo
  CC       src/unix/libuv_la-proctitle.lo
  CCLD     libuv.la
/Applications/Xcode.app/Contents/Developer/usr/bin/make  test/run-tests
  CC       test/test_run_tests-blackhole-server.o
  CC       test/test_run_tests-dns-server.o
  CC       test/test_run_tests-echo-server.o
  CC       test/test_run_tests-run-tests.o
  CC       test/test_run_tests-runner.o
  CC       test/test_run_tests-test-active.o
  CC       test/test_run_tests-test-async.o
  CC       test/test_run_tests-test-async-null-cb.o
  CC       test/test_run_tests-test-barrier.o
  CC       test/test_run_tests-test-callback-order.o
  CC       test/test_run_tests-test-callback-stack.o
  CC       test/test_run_tests-test-close-fd.o
  CC       test/test_run_tests-test-close-order.o
  CC       test/test_run_tests-test-condvar.o
  CC       test/test_run_tests-test-connection-fail.o
  CC       test/test_run_tests-test-cwd-and-chdir.o
  CC       test/test_run_tests-test-default-loop-close.o
  CC       test/test_run_tests-test-delayed-accept.o
  CC       test/test_run_tests-test-dlerror.o
  CC       test/test_run_tests-test-embed.o
  CC       test/test_run_tests-test-emfile.o
  CC       test/test_run_tests-test-error.o
  CC       test/test_run_tests-test-fail-always.o
  CC       test/test_run_tests-test-fs-event.o
  CC       test/test_run_tests-test-fs-poll.o
  CC       test/test_run_tests-test-fs.o
  CC       test/test_run_tests-test-get-currentexe.o
  CC       test/test_run_tests-test-get-loadavg.o
  CC       test/test_run_tests-test-get-memory.o
  CC       test/test_run_tests-test-getaddrinfo.o
  CC       test/test_run_tests-test-getnameinfo.o
  CC       test/test_run_tests-test-getsockname.o
  CC       test/test_run_tests-test-handle-fileno.o
  CC       test/test_run_tests-test-hrtime.o
  CC       test/test_run_tests-test-idle.o
  CC       test/test_run_tests-test-ip4-addr.o
  CC       test/test_run_tests-test-ip6-addr.o
  CC       test/test_run_tests-test-ipc-send-recv.o
  CC       test/test_run_tests-test-ipc.o
  CC       test/test_run_tests-test-loop-handles.o
  CC       test/test_run_tests-test-loop-alive.o
  CC       test/test_run_tests-test-loop-close.o
  CC       test/test_run_tests-test-loop-stop.o
  CC       test/test_run_tests-test-loop-time.o
  CC       test/test_run_tests-test-multiple-listen.o
  CC       test/test_run_tests-test-mutexes.o
  CC       test/test_run_tests-test-osx-select.o
  CC       test/test_run_tests-test-pass-always.o
  CC       test/test_run_tests-test-ping-pong.o
  CC       test/test_run_tests-test-pipe-bind-error.o
  CC       test/test_run_tests-test-pipe-connect-error.o
  CC       test/test_run_tests-test-pipe-getsockname.o
  CC       test/test_run_tests-test-pipe-sendmsg.o
  CC       test/test_run_tests-test-pipe-server-close.o
  CC       test/test_run_tests-test-pipe-close-stdout-read-stdin.o
  CC       test/test_run_tests-test-platform-output.o
  CC       test/test_run_tests-test-poll-close.o
  CC       test/test_run_tests-test-poll-closesocket.o
  CC       test/test_run_tests-test-poll.o
  CC       test/test_run_tests-test-process-title.o
  CC       test/test_run_tests-test-ref.o
  CC       test/test_run_tests-test-run-nowait.o
  CC       test/test_run_tests-test-run-once.o
  CC       test/test_run_tests-test-semaphore.o
  CC       test/test_run_tests-test-shutdown-close.o
  CC       test/test_run_tests-test-shutdown-eof.o
  CC       test/test_run_tests-test-shutdown-twice.o
  CC       test/test_run_tests-test-signal-multiple-loops.o
  CC       test/test_run_tests-test-signal.o
  CC       test/test_run_tests-test-socket-buffer-size.o
  CC       test/test_run_tests-test-spawn.o
  CC       test/test_run_tests-test-stdio-over-pipes.o
  CC       test/test_run_tests-test-tcp-bind-error.o
  CC       test/test_run_tests-test-tcp-bind6-error.o
  CC       test/test_run_tests-test-tcp-close-accept.o
  CC       test/test_run_tests-test-tcp-close-while-connecting.o
  CC       test/test_run_tests-test-tcp-close.o
  CC       test/test_run_tests-test-tcp-connect-error-after-write.o
  CC       test/test_run_tests-test-tcp-connect-error.o
  CC       test/test_run_tests-test-tcp-connect-timeout.o
  CC       test/test_run_tests-test-tcp-connect6-error.o
  CC       test/test_run_tests-test-tcp-flags.o
  CC       test/test_run_tests-test-tcp-open.o
  CC       test/test_run_tests-test-tcp-read-stop.o
  CC       test/test_run_tests-test-tcp-shutdown-after-write.o
  CC       test/test_run_tests-test-tcp-unexpected-read.o
  CC       test/test_run_tests-test-tcp-write-to-half-open-connection.o
  CC       test/test_run_tests-test-tcp-write-after-connect.o
  CC       test/test_run_tests-test-tcp-writealot.o
  CC       test/test_run_tests-test-tcp-try-write.o
  CC       test/test_run_tests-test-tcp-write-queue-order.o
  CC       test/test_run_tests-test-thread-equal.o
  CC       test/test_run_tests-test-thread.o
  CC       test/test_run_tests-test-threadpool-cancel.o
  CC       test/test_run_tests-test-threadpool.o
  CC       test/test_run_tests-test-timer-again.o
  CC       test/test_run_tests-test-timer-from-check.o
  CC       test/test_run_tests-test-timer.o
  CC       test/test_run_tests-test-tty.o
  CC       test/test_run_tests-test-udp-bind.o
  CC       test/test_run_tests-test-udp-dgram-too-big.o
  CC       test/test_run_tests-test-udp-ipv6.o
  CC       test/test_run_tests-test-udp-multicast-interface.o
  CC       test/test_run_tests-test-udp-multicast-interface6.o
  CC       test/test_run_tests-test-udp-multicast-join.o
  CC       test/test_run_tests-test-udp-multicast-join6.o
  CC       test/test_run_tests-test-udp-multicast-ttl.o
  CC       test/test_run_tests-test-udp-open.o
  CC       test/test_run_tests-test-udp-options.o
  CC       test/test_run_tests-test-udp-send-and-recv.o
  CC       test/test_run_tests-test-udp-send-immediate.o
  CC       test/test_run_tests-test-udp-send-unreachable.o
  CC       test/test_run_tests-test-udp-try-send.o
  CC       test/test_run_tests-test-walk-handles.o
  CC       test/test_run_tests-test-watcher-cross-stop.o
  CC       test/test_run_tests-runner-unix.o
  CCLD     test/run-tests
/Applications/Xcode.app/Contents/Developer/usr/bin/make  check-TESTS
[%   0|+   0|-   0|T   0|S   0]: platform_output
Output from process `platform_output`:
uv_get_process_title: /Users/hems/git/loopcast/appcast/Appcast/deps/libuv/test/.libs/run-tests
uv_resident_set_memory: 593920
uv_uptime: 416572.000000
uv_getrusage:
  user: 0 sec 972 microsec
  system: 0 sec 1109 microsec
uv_cpu_info:
  model: Intel(R) Core(TM) i7-2635QM CPU @ 2.00GHz
  speed: 2000
  times.sys: 35976870
  times.user: 69336390
  times.idle: 165557690
  times.irq: 0
  times.nice: 0
  model: Intel(R) Core(TM) i7-2635QM CPU @ 2.00GHz
  speed: 2000
  times.sys: 6551090
  times.user: 42575950
  times.idle: 221250190
  times.irq: 0
  times.nice: 0
  model: Intel(R) Core(TM) i7-2635QM CPU @ 2.00GHz
  speed: 2000
  times.sys: 17615010
  times.user: 56376690
  times.idle: 196385620
  times.irq: 0
  times.nice: 0
  model: Intel(R) Core(TM) i7-2635QM CPU @ 2.00GHz
  speed: 2000
  times.sys: 4817910
  times.user: 40160400
  times.idle: 225398880
  times.irq: 0
  times.nice: 0
  model: Intel(R) Core(TM) i7-2635QM CPU @ 2.00GHz
  speed: 2000
  times.sys: 15460210
  times.user: 52943520
  times.idle: 201973560
  times.irq: 0
  times.nice: 0
  model: Intel(R) Core(TM) i7-2635QM CPU @ 2.00GHz
  speed: 2000
  times.sys: 4316560
  times.user: 39211610
  times.idle: 226849000
  times.irq: 0
  times.nice: 0
  model: Intel(R) Core(TM) i7-2635QM CPU @ 2.00GHz
  speed: 2000
  times.sys: 14397410
  times.user: 51247620
  times.idle: 204732230
  times.irq: 0
  times.nice: 0
  model: Intel(R) Core(TM) i7-2635QM CPU @ 2.00GHz
  speed: 2000
  times.sys: 4071700
  times.user: 38768140
  times.idle: 227537290
  times.irq: 0
  times.nice: 0
uv_interface_addresses:
  name: lo0
  internal: 1
  physical address: 00:00:00:00:00:00
  address: ::1
  netmask: ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff
  name: lo0
  internal: 1
  physical address: 00:00:00:00:00:00
  address: 127.0.0.1
  netmask: 255.0.0.0
  name: lo0
  internal: 1
  physical address: 00:00:00:00:00:00
  address: fe80::1
  netmask: ffff:ffff:ffff:ffff::
  name: en1
  internal: 0
  physical address: e0:f8:47:43:8f:6c
  address: fe80::e2f8:47ff:fe43:8f6c
  netmask: ffff:ffff:ffff:ffff::
  name: en1
  internal: 0
  physical address: e0:f8:47:43:8f:6c
  address: 192.168.0.2
  netmask: 255.255.255.0
=============================================================
[%  72|+ 177|-   0|T   0|S   0]: spawn_reads_child_path
`spawn_reads_child_path` failed: exit code 6
Output from process `spawn_reads_child_path`:
exit_cb
Assertion failed in test/test-spawn.c on line 66: exit_status == 1
=============================================================
[%  77|+ 188|-   1|T   0|S   0]: osx_selectgot some input
with a couple of lines
feel pretty happy
[%  77|+ 189|-   1|T   0|S   0]: osx_select_many_fdsgot some input
with a couple of lines
feel pretty happy
[%  86|+ 211|-   1|T   0|S   0]: fs_event_watch_dir
`fs_event_watch_dir` failed: timeout
Output from process `fs_event_watch_dir`: (no output)
=============================================================
[% 100|+ 243|-   2|T   0|S   0]: Done.
FAIL: test/run-tests
======================================================
1 of 1 test failed
Please report to https://github.com/libuv/libuv/issues
======================================================
make[1]: *** [check-TESTS] Error 1
make: *** [check-am] Error 2
 ./install-sh -c -d '/usr/local/lib'
 /bin/sh ./libtool   --mode=install /usr/bin/install -c   libuv.la '/usr/local/lib'
libtool: install: /usr/bin/install -c .libs/libuv.1.dylib /usr/local/lib/libuv.1.dylib
libtool: install: (cd /usr/local/lib && { ln -s -f libuv.1.dylib libuv.dylib || { rm -f libuv.dylib && ln -s libuv.1.dylib libuv.dylib; }; })
libtool: install: /usr/bin/install -c .libs/libuv.lai /usr/local/lib/libuv.la
libtool: install: /usr/bin/install -c .libs/libuv.a /usr/local/lib/libuv.a
libtool: install: chmod 644 /usr/local/lib/libuv.a
libtool: install: ranlib /usr/local/lib/libuv.a
 ./install-sh -c -d '/usr/local/include'
 /usr/bin/install -c -m 644 include/uv.h include/uv-errno.h include/uv-threadpool.h include/uv-version.h include/uv-unix.h include/uv-darwin.h '/usr/local/include'
 ./install-sh -c -d '/usr/local/lib/pkgconfig'
 /usr/bin/install -c -m 644 libuv.pc '/usr/local/lib/pkgconfig'
deps/libuv [                                                                                                                           master ] 2:57 pm
````

**Comment by vielmetti** [comment]
I get the same failure on OS X 10.9.5.

**Comment by whitlockjc** [comment]
I just started looking into this and I was able to reproduce the flaky test failure on OS X but the failure is slightly different, but similar enough to report here.  I am on OS X 10.10.4  and I ran the test suite a handful of times without a test failure.  Then ran it a sixth time and `fs_event_watch_dir_recursive` failed.  As I look into this, I will update you with anything I find.

**Comment by iankronquist** [comment]
@whitlockjc Good luck! Some people think it's a race condition.

**Comment by whitlockjc** [comment]
Can someone apply the `osx` label?  Also, what are the chances of getting a `flaky-test` label like in [nodejs/labels/flaky-test](https://github.com/joyent/node/labels/flaky-test)?

**Comment by saghul** [comment]
@whitlockjc Thanks for looking into this! I applied to "osx" label. We don't use "flaky-test" in libuv,it either passes, or it's a bug, like in this case :-)

**Comment by whitlockjc** [comment]
Works for me.

**Comment by whitlockjc** [comment]
Well, I've got some good news, I've figured out the cause of the race condition.  `fs_event_watch_dir` uses `fs_event_create_files` to create some files, which drives the tests, and then schedules a timer to cleanup after itself.  What can happen is you can end up with the cleanup callback *(`fs_event_unlink_files`)* running prior to all of the `fs_event` handler callbacks *(as a result of the files being created)* being processed.  This is not the only race condition.  The cleanup callback itself removes files, creating more `fs_event`, and then schedules another timer whose callback *(`fs_event_create_files`)* can run prior to all of the removal events being processed.

This will leave some `fs_event` handlers not having completed which keeps the event loop open, hence the timeout.  `fs_event_watch_dir_recursive` is also flaky and happens to use the same testing logic so I think we should be able to fix both flaky tests in one pass using the same approach.

My plan is to refactor the tests to not use timers so this can be avoided.  Once I've got this working, I will submit a PR.  The way I validated this was keeping track of each of these callbacks, their order and such.  I noticed that when things failed, not all `fs_event` handlers were processed.  I bumped up the timers involved to use a timeout of `200ms` instead of `50ms` and the failures went away.  I can run both flaky tests 1000 times without failure, where I use to run somewhere around 10 before I got the first failure.  Below is the diff of my current changes which show this is clearly a timing issue:

```diff
diff --git a/test/test-fs-event.c b/test/test-fs-event.c
index f6c2214..245e550 100644
--- a/test/test-fs-event.c
+++ b/test/test-fs-event.c
@@ -195,7 +195,7 @@ static void fs_event_create_files(uv_timer_t* handle) {
     create_file(handle->loop, fs_event_get_filename(i));
 
   /* And unlink them */
-  ASSERT(0 == uv_timer_start(&timer, fs_event_unlink_files, 50, 0));
+  ASSERT(0 == uv_timer_start(&timer, fs_event_unlink_files, 200, 0));
 }
 
 static void fs_event_create_files_in_subdir(uv_timer_t* handle) {
@@ -212,7 +212,7 @@ static void fs_event_create_files_in_subdir(uv_timer_t* handle) {
     create_file(handle->loop, fs_event_get_filename_in_subdir(i));
 
   /* And unlink them */
-  ASSERT(0 == uv_timer_start(&timer, fs_event_unlink_files_in_subdir, 50, 0));
+  ASSERT(0 == uv_timer_start(&timer, fs_event_unlink_files_in_subdir, 200, 0));
 }
 
 void fs_event_unlink_files_in_subdir(uv_timer_t* handle) {
@@ -230,7 +230,7 @@ void fs_event_unlink_files_in_subdir(uv_timer_t* handle) {
 
   /* And create them again */
   if (handle != NULL)
-    ASSERT(0 == uv_timer_start(&timer, fs_event_create_files_in_subdir, 50, 0));
+    ASSERT(0 == uv_timer_start(&timer, fs_event_create_files_in_subdir, 200, 0));
 }
 
 void fs_event_unlink_files(uv_timer_t* handle) {
@@ -248,7 +248,7 @@ void fs_event_unlink_files(uv_timer_t* handle) {
 
   /* And create them again */
   if (handle != NULL)
-    ASSERT(0 == uv_timer_start(&timer, fs_event_create_files, 50, 0));
+    ASSERT(0 == uv_timer_start(&timer, fs_event_create_files, 200, 0));
 }
 
 static void fs_event_cb_file(uv_fs_event_t* handle, const char* filename,
```

/cc @indutny

**Comment by saghul** [comment]
> What can happen is you can end up with the cleanup callback (fs_event_unlink_files) running prior to all of the fs_event handler callbacks (as a result of the files being created) being processed

How is this possible, if removing the files always happens 50ms after they where all created? I guess it's possible that if you create a file and remove it in less than 50ms you lose the event?

For this particular test, maybe we shouldn't have `fs_event_unlink_files` create more files after they have been removed.

> My plan is to refactor the tests to not use timers so this can be avoided. Once I've got this working, I will submit a PR. The way I validated this was keeping track of each of these callbacks, their order and such.

Sounds like a plan! Thanks a lot for your help tracking this down!

**Comment by whitlockjc** [comment]
No problem.  As for your first question, you're right.  What would happen is we'd create a file and remove it within 50ms and then we'd recreate it so the original handle was still open.  It took a lot of debugging to find this but that's definitely the situation I was able to reproduce.

**Comment by saghul** [comment]
Good to know! Please add a note about this to the documentation when you
make the patch. Thanks!

On Fri, Aug 14, 2015 at 5:39 PM, Jeremy Whitlock <notifications@github.com>
wrote:

> No problem. As for your first question, you're right. What would happen is
> we'd create a file and remove it within 50ms and then we'd recreate it so
> the original handle was still open. It took a lot of debugging to find this
> but that's definitely the situation I was able to reproduce.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/libuv/libuv/issues/30#issuecomment-131153715>.
>



-- 
/Saúl
bettercallsaghul.com

**Comment by whitlockjc** [comment]
I've got a PR ready.  What should the subsystem be?  I'd figure `tests` since I'm only changing the tests but since this is related to `fs` events, it could be `fs` as well.  Also, what documentation are you referring to?

**Comment by saghul** [comment]
"test,fs" will do.

**Comment by saghul** [comment]
Fixed by #480.
